### ISSUE TITLE
fit_generator reporting incorrect estimated time to complete

### ISSUE THREAD

**Issue by jetojedno** [keras-team/keras#9748]
when using fit_generator with a validation generator, the progress gets to the penultimate training batch & stops, and the time freezes at that point for the period needed to process the validation batches.

This is in jupyter notebook (IPython = 6.2.1), keras 2.1.5 & tensorflow 1.4.1.

It would be helpful if the batch number got to the final batch :-) and then something indicated that validation is happening.
Also, the estimated time to complete should include an allowance for validation batches.  It would be reasonable to allow half the time of a training batch until the first epoch, and then update the multiplier.

Thank you!

- [ ] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/keras-team/keras.git --upgrade --no-deps

- [ ] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [ ] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short). - I'd also need to supply all the training data etc.

Here is the output I've been looking at for the past 3 minutes:
```
Epoch 1/10
74/74 [==============================] - 1308s 18s/step - loss: 9.3432 - acc: 0.5959 - val_loss: 8.2431 - val_acc: 0.6018

Epoch 00001: val_loss improved from inf to 8.24310, saving model to checkpoints/dermatology_v1_I-20180324_230205/best.ckpt
Epoch 2/10
74/74 [==============================] - 1254s 17s/step - loss: 7.5390 - acc: 0.6018 - val_loss: 7.0590 - val_acc: 0.6018

Epoch 00002: val_loss improved from 8.24310 to 7.05902, saving model to checkpoints/dermatology_v1_I-20180324_230205/best.ckpt
Epoch 3/10
73/74 [============================>.] - ETA: 7s - loss: 6.8114 - acc: 0.6036
``` 

I cannot update the software versions easily as I'm part way through a course, and I need to stick with the versions they're using. :-(

**Comment by tRosenflanz** [comment]
I would add timers for callbacks too - I do validation from the callback since I am getting a little bit more specific metrics than loss/accuracy.

**Comment by jetojedno** [comment]
@tRosenflanz do you have a suggestion for how the estimated time to complete is calculated for each part?
 - validation is (approximately) a sum of linear products of number of validation batches and number of validation samples (multipliers are the estimator factors) plus a constant time (another estimator factor)  
 - weights / checkpoints saving is a constant time  
... would those three factors cover most situations?

**Comment by tRosenflanz** [comment]
I am not sure if time to completion estimation is feasible given that you cannot really know how long each arbitrary code in callback will really take in advance. However, I think it is a reasonable assumption to say that it will take approximately the same time across each epoch so the information from epoch 1 completion could be used as an approximation for further epochs

**Comment by jetojedno** [comment]
That's a better suggestion than mine. Simpler.

**Comment by jeromemassot** [comment]
I have the same issue here, and I was thinking it was due to PlaidML as I am using Radeon GPU. But in fact I guess that you are using NVidia and still experiencing this dramatic decrease of performance when using fit_generator and validation. I am not experienced enough to propose a fix so I will just wait for the next update :)

**Comment by jetojedno** [comment]
Yes, I have Nvidia GPUs.  The decrease, in my case, is coming from the generator which is doing a lot of work enhancing the images.  It's using all the CPUs and the model is running on the GPUs.

**Comment by singhay** [comment]
+1

**Comment by marisn** [comment]
For a starter the progress bar could count number of batches correctly. When training is complete and only validation part is left, the progress bar indicates only one batch to go, but in the reality there could be tens of batches remaining. Here is an example of a model.fit_generator run with validation batch generator:
```
Epoch 1/100
157/158 [============================>.] - ETA: 4s - loss: 0.6937 - acc: 0.5069 1 validation batch
 2 validation batch
...
 59 validation batch
 60 validation batch
```
Also – in the example it displays 4s remaining, but in the reality it will take more than 20 minutes.
One of options could be providing a separate progress bar for validation part of epoch. Although I would prefer to have single progress bar to `rule them all`.
